src/Microsoft.AspNetCore.Http.Abstractions/PathString.cs

```diff
@@ -71,20 +72,45 @@ public string ToUriComponent()
             {
                 return string.Empty;
             }
-            var values = _value.Split(splitChar);
-            var changed = false;
-            for (var i = 0; i < values.Length; i++)
-            {
-                var value = values[i];
-                values[i] = UrlEncoder.Default.Encode(value);
 
-                if (!changed && value != values[i])
+            StringBuilder buffer = null;
+
+            var start = 0;
+            var count = 0;
+            for (int i = 0; i < _value.Length; ++i)
+            {
+                if (IsLegalPathChar(_value[i]))
+                {
+                    count++;
+                }
+                else
                 {
-                    changed = true;
+                    if (buffer == null)
+                    {
+                        buffer = new StringBuilder();
+                    }
+
+                    // wrap up previous chars
+                    buffer.Append(_value.Substring(start, count));
+
+                    // escape the string
+                    buffer.Append(Uri.EscapeDataString(_value.Substring(i, 1)));
+
+                    // reset pointer
+                    start = i + 1;
+                    count = 0;
                 }
             }
 
-            return changed ? string.Join("/", values) : _value;
+            if (count == _value.Length)
+            {
+                return _value;
+            }
+            else
```

Comment:
Actually doesn't matter, if `count == 0` the following `substring` won't generate result. In both case the buffer needs to be returned. To save a needless append operation, I'll change it to:

* `count > 0` => append remaining unescaped chars, and then return buffer.
* `count == 0` (else) => return buffer.